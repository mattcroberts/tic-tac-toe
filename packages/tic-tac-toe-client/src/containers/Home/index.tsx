import * as React from "react";
import { graphql } from "react-apollo";
import { withRouter } from "react-router-dom";
import { compose } from "recompose";
import NEW_GAME from "./newGame.graphql";

interface IProps {
    newGame: () => undefined;
}

class Home extends React.Component<IProps> {
    constructor(props: IProps) {
        super(props);
        this.onClick = this.onClick.bind(this);
    }

    public async onClick() {
        const { history, newGame }: any = this.props;
        const {
            data: { newGame: grid }
        } = await newGame();

        if (grid.id) {
            history.push(`/game/${grid.id}`);
        }
    }

    public render() {
        return (
            <React.Fragment>
                <button onClick={this.onClick}>Play</button>
            </React.Fragment>
        );
    }
}

export default compose(
    withRouter,
    graphql(NEW_GAME, {
        name: "newGame"
    })
)(Home);
