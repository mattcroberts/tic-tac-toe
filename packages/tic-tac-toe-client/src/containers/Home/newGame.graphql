mutation newGame {
    newGame {
        id
        gridItems {
            id
            player {
                id,
                symbol
            }
        }
        currentPlayer {
            id
            symbol
        }
        isFinished
    }
}
