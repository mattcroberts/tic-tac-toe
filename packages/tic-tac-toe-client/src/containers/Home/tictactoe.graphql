query tictactoe {
    tictactoe {
        gamesInProgress
        gamesFinished
        gamesDrawn
        naughtWins
        crossWins
    }
}
