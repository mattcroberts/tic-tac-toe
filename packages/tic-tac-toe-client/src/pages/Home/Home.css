:root {
    --heading-color: #000;
    --button-border-color: #00b5ec;
    --button-color: #00b5ec;
    --button-text-color: #fff;
}

.root {
    display: flex;
    flex-direction: column;
    align-items: center;
}

.root > * {
    flex: 1 1;
    display: flex;
}

.root > *:first-child {
    flex-direction: column-reverse;
}

.root > *:nth-child(2) {
    flex-direction: column;
}

@keyframes glow {
    from {
        text-shadow: 0 0 0 var(--heading-color);
    }

    to {
        text-shadow: 0 0 20px var(--heading-color);
    }
}

.heading {
    font-size: 42px;
    font-weight: 800;
    color: var(--heading-color);
}

.heading.glow {
    animation-duration: 1s;
    animation-name: glow;
    animation-iteration-count: infinite;
    animation-direction: alternate;
}

.button {
    color: var(--button-text-color);
    border: 1px solid var(--button-border-color);
    border-radius: 3px;
    background-color: var(--button-color);
    margin-right: 20px;
}

.button:last-child {
    margin-right: 0;
}

.buttonContainer {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    margin-bottom: 20px;
}

.button:disabled {
    background-color: var(--button-border-color);
}

.stats {
    background-color: var(--button-color);
    padding: 6px;
    color: var(--button-text-color);
    border: 1px solid var(--button-border-color);
    border-radius: 3px;
}

.cell {
    padding: 3px;
}
