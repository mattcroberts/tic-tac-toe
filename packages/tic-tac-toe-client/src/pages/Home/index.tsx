import * as React from "react";
import classnames from "classnames";

import Button from "../../components/Button";

import style from "./Home.css";
import { TicTacToe } from "../../../typings/types";

const renderStats = (stats: TicTacToe) => {
    if (!stats) {
        return null;
    }

    return (
        <table className={style.stats}>
            <tbody>
                <tr>
                    <td className={style.cell}>Games In Progress</td>
                    <td className={style.cell}>{stats.gamesInProgress}</td>
                </tr>
                <tr>
                    <td className={style.cell}>Games Finished</td>
                    <td className={style.cell}>{stats.gamesFinished}</td>
                </tr>
                <tr>
                    <td className={style.cell}>Naught Wins</td>
                    <td className={style.cell}>{stats.naughtWins}</td>
                </tr>
                <tr>
                    <td className={style.cell}>Cross Wins</td>
                    <td className={style.cell}>{stats.crossWins}</td>
                </tr>
                <tr>
                    <td className={style.cell}>Games Drawn</td>
                    <td className={style.cell}>{stats.gamesDrawn}</td>
                </tr>
            </tbody>
        </table>
    );
};

const Home = ({
    loading = false,
    newGame,
    gameInfo
}: {
    loading: boolean;
    newGame: () => any;
    gameInfo: TicTacToe;
}) => (
    <div className={style.root}>
        <div>
            <h1
                className={classnames(style.heading, { [style.glow]: loading })}
            >
                Tic Tac Toe
            </h1>
        </div>
        <div>
            <div className={style.buttonContainer}>
                <Button
                    className={style.button}
                    disabled={loading}
                    onClick={newGame}
                >
                    {"New Game"}
                </Button>
                <Button
                    className={style.button}
                    disabled={loading}
                    onClick={newGame}
                >
                    {"Multiplayer"}
                </Button>
            </div>
            {renderStats(gameInfo)}
        </div>
    </div>
);

export default Home;
