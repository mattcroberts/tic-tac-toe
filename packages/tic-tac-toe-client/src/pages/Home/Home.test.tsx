import * as React from "react";
import { shallow } from "enzyme";
import Home from ".";

describe("Home", () => {
    const defaultProps = {
        children: <div>Test</div>,
        loading: false,
        newGame: jest.fn(),
        gameInfo: {
            gamesInProgress: 0,
            gamesFinished: 0,
            gamesDrawn: 0,
            naughtWins: 0,
            crossWins: 0
        }
    };

    beforeEach(() => {
        defaultProps.newGame.mockClear();
    });

    it("should render", () => {
        const component = shallow(<Home {...defaultProps} />);
        expect(component).toMatchSnapshot();
    });

    it("should render loading indicator", () => {
        const component = shallow(<Home {...defaultProps} loading={true} />);
        expect(component).toMatchSnapshot();
    });
});
