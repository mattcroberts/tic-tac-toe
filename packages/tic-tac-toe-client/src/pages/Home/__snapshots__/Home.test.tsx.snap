// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should render 1`] = `
<div
  className="root"
>
  <div>
    <h1
      className="heading"
    >
      Tic Tac Toe
    </h1>
  </div>
  <div>
    <div
      className="buttonContainer"
    >
      <Button
        className="button"
        disabled={false}
        onClick={[Function]}
      >
        New Game
      </Button>
      <Button
        className="button"
        disabled={false}
        onClick={[Function]}
      >
        Multiplayer
      </Button>
    </div>
    <table
      className="stats"
    >
      <tbody>
        <tr>
          <td
            className="cell"
          >
            Games In Progress
          </td>
          <td
            className="cell"
          >
            0
          </td>
        </tr>
        <tr>
          <td
            className="cell"
          >
            Games Finished
          </td>
          <td
            className="cell"
          >
            0
          </td>
        </tr>
        <tr>
          <td
            className="cell"
          >
            Naught Wins
          </td>
          <td
            className="cell"
          >
            0
          </td>
        </tr>
        <tr>
          <td
            className="cell"
          >
            Cross Wins
          </td>
          <td
            className="cell"
          >
            0
          </td>
        </tr>
        <tr>
          <td
            className="cell"
          >
            Games Drawn
          </td>
          <td
            className="cell"
          >
            0
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`Home should render loading indicator 1`] = `
<div
  className="root"
>
  <div>
    <h1
      className="heading glow"
    >
      Tic Tac Toe
    </h1>
  </div>
  <div>
    <div
      className="buttonContainer"
    >
      <Button
        className="button"
        disabled={true}
        onClick={[Function]}
      >
        New Game
      </Button>
      <Button
        className="button"
        disabled={true}
        onClick={[Function]}
      >
        Multiplayer
      </Button>
    </div>
    <table
      className="stats"
    >
      <tbody>
        <tr>
          <td
            className="cell"
          >
            Games In Progress
          </td>
          <td
            className="cell"
          >
            0
          </td>
        </tr>
        <tr>
          <td
            className="cell"
          >
            Games Finished
          </td>
          <td
            className="cell"
          >
            0
          </td>
        </tr>
        <tr>
          <td
            className="cell"
          >
            Naught Wins
          </td>
          <td
            className="cell"
          >
            0
          </td>
        </tr>
        <tr>
          <td
            className="cell"
          >
            Cross Wins
          </td>
          <td
            className="cell"
          >
            0
          </td>
        </tr>
        <tr>
          <td
            className="cell"
          >
            Games Drawn
          </td>
          <td
            className="cell"
          >
            0
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
